### PR TITLE
Move TypeSerializer and DataSerializer to `core` module (part 1)

### PR DESCRIPTION
**Pull request by aslesarenko**
In this PR:
- [x] SType hierarchy moved to `core` module `sigma.ast` package
- [x] all SMethod setup is moved to MethodsContainer hierarchy

### CODE REVIEW

**Review by kushti — APPROVED** [review]

File: core/shared/src/main/scala/sigma/ast/SType.scala [hunk 1]
```diff
@@ -0,0 +1,853 @@
...
+import sigma.ast.SCollection.SByteArray
+import sigma.ast.SType.TypeCode
+import sigma.data.OverloadHack.Overloaded1
+import sigma.data.{CBigInt, Nullable, RType, SigmaConstants}
+import sigma.reflection.{RClass, RMethod, ReflectionData}
+import sigma.util.Extensions.{IntOps, LongOps, ShortOps}
+import sigma.{AvlTree, BigInt, Box, Coll, Colls, Context, Evaluation, GroupElement, Header, PreHeader, SigmaDslBuilder, SigmaProp}
+
+import java.math.BigInteger
+import scala.util.{Failure, Success}
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply core/shared/src/main/scala/sigma/ast/SType.scala:14]: fixed

File: core/shared/src/main/scala/sigma/ast/SType.scala [hunk 2]
```diff
@@ -0,0 +1,853 @@
...
+
+/** Base type for all AST nodes of sigma lang. */
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply core/shared/src/main/scala/sigma/ast/SType.scala:16]: fixed

File: core/shared/src/main/scala/sigma/ast/SType.scala [hunk 3]
```diff
@@ -0,0 +1,853 @@
...
+    val t = Evaluation.stypeToRType(this)
+    t.name
+  }
+}
+
+object SType {
+  /** Representation of type codes used in serialization. */
+  type TypeCode = Byte
+
+  val DummyValue = 0.asWrappedType
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply core/shared/src/main/scala/sigma/ast/SType.scala:69]: fixed

File: core/shared/src/main/scala/sigma/ast/SType.scala [hunk 4]
```diff
@@ -0,0 +1,853 @@
...
+}
+
+object SCollection extends STypeCompanion {
+  override val reprClass: RClass[_] = RClass(classOf[Coll[_]])
+  override def typeId = SCollectionType.CollectionTypeCode
+
+  /** Costructs a collection type with the given type of elements. */
+  implicit def typeCollection[V <: SType](implicit tV: V): SCollection[V] = SCollection[V](tV)
+
+  import SType.{paramIV, paramIVSeq, paramOV}
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply core/shared/src/main/scala/sigma/ast/SType.scala:698]: fixed

File: core/shared/src/main/scala/sigma/ast/STypeParam.scala [hunk 5]
```diff
@@ -0,0 +1,14 @@
+package sigma.ast
+
+case class STypeParam(
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply core/shared/src/main/scala/sigma/ast/STypeParam.scala:3]: done

**Pull request merged**